File: app/src/main/java/com/amaury/pointage/v2/engine/EmployerUnemploymentAgsV2.kt
package com.amaury.pointage.v2.engine

import java.time.YearMonth
import kotlin.math.min

/**
 * Assurance chômage + AGS employeur.
 * Les taux sont confirmés par entreprise/période : aucun taux de droit commun
 * n'est imposé automatiquement car bonus-malus et cas ETT peuvent s'appliquer.
 */
object EmployerUnemploymentAgsV2 {
    data class Record(
        val id: String,
        /** Taux décimal : 4,00 % = 0,04. */
        val unemploymentRate: Double,
        /** Taux décimal : 0,25 % = 0,0025. */
        val agsRate: Double,
        val effectiveFrom: YearMonth,
        val effectiveTo: YearMonth? = null,
        val source: String
    )

    data class Snapshot(
        val unemploymentRate: Double?,
        val agsRate: Double?,
        val source: String?,
        val reliable: Boolean,
        val warnings: List<String>
    )

    data class Result(
        val baseAmount: Double?,
        val unemploymentAmount: Double?,
        val agsAmount: Double?,
        val totalEmployerAmount: Double?,
        val complete: Boolean,
        val warnings: List<String>
    )

    fun resolve(records: List<Record>, period: YearMonth): Snapshot {
        val malformed = records.filter { record ->
            record.source.isBlank() ||
                record.effectiveTo?.let { it < record.effectiveFrom } == true ||
                !validRate(record.unemploymentRate) ||
                !validRate(record.agsRate)
        }
        if (malformed.isNotEmpty()) {
            return Snapshot(
                unemploymentRate = null,
                agsRate = null,
                source = null,
                reliable = false,
                warnings = listOf("Chômage/AGS employeur : une règle enregistrée est incomplète ou incohérente ; aucun taux n'est appliqué.")
            )
        }

        val active = records.filter { record ->
            period >= record.effectiveFrom && (record.effectiveTo == null || period <= record.effectiveTo)
        }
        if (active.isEmpty()) {
            return Snapshot(
                unemploymentRate = null,
                agsRate = null,
                source = null,
                reliable = false,
                warnings = listOf("Chômage/AGS employeur : taux confirmés à renseigner pour ${period.monthValue.toString().padStart(2, '0')}/${period.year}.")
            )
        }
        if (active.size > 1) {
            return Snapshot(
                unemploymentRate = null,
                agsRate = null,
                source = null,
                reliable = false,
                warnings = listOf("Chômage/AGS employeur : plusieurs règles se chevauchent sur la période ; calcul patronal bloqué.")
            )
        }

        val selected = active.single()
        return Snapshot(
            unemploymentRate = selected.unemploymentRate,
            agsRate = selected.agsRate,
            source = selected.source,
            reliable = true,
            warnings = emptyList()
        )
    }

    fun calculate(
        grossSocial: Double,
        fourTimesApplicableCeiling: Double?,
        unemploymentRate: Double?,
        agsRate: Double?
    ): Result {
        if (fourTimesApplicableCeiling == null || !fourTimesApplicableCeiling.isFinite() || fourTimesApplicableCeiling < 0.0) {
            return Result(null, null, null, null, false, listOf("Chômage/AGS employeur : plafond social applicable indisponible ; coût employeur incomplet."))
        }
        if (unemploymentRate == null || agsRate == null) {
            return Result(null, null, null, null, false, listOf("Chômage/AGS employeur : taux confirmés manquants ; coût employeur incomplet."))
        }
        if (!validRate(unemploymentRate) || !validRate(agsRate)) {
            return Result(null, null, null, null, false, listOf("Chômage/AGS employeur : taux invalide ; aucun montant patronal n'est calculé."))
        }

        val base = min(grossSocial.coerceAtLeast(0.0), fourTimesApplicableCeiling)
        val unemployment = base * unemploymentRate
        val ags = base * agsRate
        return Result(
            baseAmount = base,
            unemploymentAmount = unemployment,
            agsAmount = ags,
            totalEmployerAmount = unemployment + ags,
            complete = true,
            warnings = emptyList()
        )
    }

    private fun validRate(rate: Double): Boolean = rate.isFinite() && rate >= 0.0 && rate <= 1.0
}

File: ios/HPTravail/HPTravail/SalaryV2/EmployerUnemploymentAgsV2.swift
import Foundation

/// Assurance chômage + AGS employeur avec taux confirmés par entreprise et période.
/// Aucun taux de droit commun n'est imposé automatiquement : bonus-malus et cas particuliers restent explicites.
enum EmployerUnemploymentAgsV2 {
    struct Period: Equatable, Comparable {
        let year: Int
        let month: Int

        var isValid: Bool { year > 0 && (1...12).contains(month) }

        static func < (lhs: Period, rhs: Period) -> Bool {
            lhs.year == rhs.year ? lhs.month < rhs.month : lhs.year < rhs.year
        }

        var label: String { String(format: "%02d/%04d", month, year) }
    }

    struct Record: Equatable {
        let id: String
        let unemploymentRate: Double
        let agsRate: Double
        let effectiveFrom: Period
        let effectiveTo: Period?
        let source: String
    }

    struct Snapshot: Equatable {
        let unemploymentRate: Double?
        let agsRate: Double?
        let source: String?
        let reliable: Bool
        let warnings: [String]
    }

    struct Result: Equatable {
        let baseAmount: Double?
        let unemploymentAmount: Double?
        let agsAmount: Double?
        let totalEmployerAmount: Double?
        let complete: Bool
        let warnings: [String]
    }

    static func resolve(records: [Record], period: Period) -> Snapshot {
        guard period.isValid else {
            return Snapshot(
                unemploymentRate: nil,
                agsRate: nil,
                source: nil,
                reliable: false,
                warnings: ["Chômage/AGS employeur : période invalide ; aucun taux n'est appliqué."]
            )
        }

        let malformed = records.filter { record in
            record.source.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
                !record.effectiveFrom.isValid ||
                (record.effectiveTo?.isValid == false) ||
                (record.effectiveTo.map { $0 < record.effectiveFrom } ?? false) ||
                !validRate(record.unemploymentRate) ||
                !validRate(record.agsRate)
        }
        guard malformed.isEmpty else {
            return Snapshot(
                unemploymentRate: nil,
                agsRate: nil,
                source: nil,
                reliable: false,
                warnings: ["Chômage/AGS employeur : une règle enregistrée est incomplète ou incohérente ; aucun taux n'est appliqué."]
            )
        }

        let active = records.filter { record in
            period >= record.effectiveFrom && (record.effectiveTo == nil || period <= record.effectiveTo!)
        }
        guard !active.isEmpty else {
            return Snapshot(
                unemploymentRate: nil,
                agsRate: nil,
                source: nil,
                reliable: false,
                warnings: ["Chômage/AGS employeur : taux confirmés à renseigner pour \(period.label)."]
            )
        }
        guard active.count == 1, let selected = active.first else {
            return Snapshot(
                unemploymentRate: nil,
                agsRate: nil,
                source: nil,
                reliable: false,
                warnings: ["Chômage/AGS employeur : plusieurs règles se chevauchent sur la période ; calcul patronal bloqué."]
            )
        }

        return Snapshot(
            unemploymentRate: selected.unemploymentRate,
            agsRate: selected.agsRate,
            source: selected.source,
            reliable: true,
            warnings: []
        )
    }

    static func calculate(
        grossSocial: Double,
        fourTimesApplicableCeiling: Double?,
        unemploymentRate: Double?,
        agsRate: Double?
    ) -> Result {
        guard grossSocial.isFinite, grossSocial >= 0 else {
            return incomplete("Chômage/AGS employeur : assiette brute invalide ; aucun montant patronal n'est calculé.")
        }
        guard let ceiling = fourTimesApplicableCeiling,
              ceiling.isFinite,
              ceiling >= 0 else {
            return incomplete("Chômage/AGS employeur : plafond social applicable indisponible ; coût employeur incomplet.")
        }
        guard let unemploymentRate, let agsRate else {
            return incomplete("Chômage/AGS employeur : taux confirmés manquants ; coût employeur incomplet.")
        }
        guard validRate(unemploymentRate), validRate(agsRate) else {
            return incomplete("Chômage/AGS employeur : taux invalide ; aucun montant patronal n'est calculé.")
        }

        let base = min(grossSocial, ceiling)
        let unemployment = base * unemploymentRate
        let ags = base * agsRate
        return Result(
            baseAmount: base,
            unemploymentAmount: unemployment,
            agsAmount: ags,
            totalEmployerAmount: unemployment + ags,
            complete: true,
            warnings: []
        )
    }

    private static func incomplete(_ warning: String) -> Result {
        Result(
            baseAmount: nil,
            unemploymentAmount: nil,
            agsAmount: nil,
            totalEmployerAmount: nil,
            complete: false,
            warnings: [warning]
        )
    }

    private static func validRate(_ rate: Double) -> Bool {
        rate.isFinite && rate >= 0 && rate <= 1
    }
}
